### PR TITLE
Задача 3. Регулярные запросы для всех пар вершин

### PR DESCRIPTION
**Pull request by IlyaMuravjov**
* **Мягкий дедлайн**: 25.09.2022, 23:59
* **Жёсткий дедлайн**: 28.09.2022, 23:59
* Полный балл: 5

## Задача

- [x] Используя [разреженные матрицы из sciPy](https://docs.scipy.org/doc/scipy/reference/sparse.html) реализовать **функцию** пересечения двух конечных автоматов через тензорное произведение.
- [x] На основе предыдущей функции реализовать **функцию** выполнения регулярных запросов к графам. Не забыть предоставить возможность задавать стартовые и финальные вершины в графе.
  - Для конструирования регулярного запроса и преобразований графа использовать результаты [Задачи 2](https://github.com/JetBrains-Research/formal-lang-course/blob/main/tasks/task2.md).
- [x] Добавить необходимые тесты.

### CODE REVIEW

File: project/bool_decomposed_nfa.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+from typing import Callable
+from typing import Dict
+from typing import Set
+from typing import Tuple
+
+from pyformlang.finite_automaton import NondeterministicFiniteAutomaton as NFA
+from pyformlang.finite_automaton import State
+from pyformlang.finite_automaton import Symbol
```
Comment (vivalerievich) [inline project/bool_decomposed_nfa.py]:
```suggestion
import pyformlang
```

File: project/bool_decomposed_nfa.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+__all__ = ["BoolDecomposedNFA"]
+
+
+class BoolDecomposedNFA:
+    state_count: int
+    idx_to_state: Callable[[int], State]
+    adj_bool_decomposition: BoolDecomposition[Symbol]
+    start_state_indices: Set[int]
+    final_state_indices: Set[int]
```
Comment (vivalerievich) [inline project/bool_decomposed_nfa.py]: Зачем Вам именно у объекта класса эти атрибуты, а не у его инстансов?

File: project/bool_decomposed_nfa.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        adj_bool_decomposition = BoolDecomposition(
+            len(nfa.states),
+            content=(
+                (state_to_idx_dict[source], state_to_idx_dict[target], symbol)
+                for (source, symbol, target) in nfa
+            ),
+        )
+        return BoolDecomposedNFA(
+            state_count=len(nfa.states),
+            idx_to_state=idx_to_state_dict.__getitem__,
```
Comment (vivalerievich) [inline project/bool_decomposed_nfa.py:46]: Это вдруг зачем?

File: project/bool_decomposition.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+from typing import Generic
+from typing import Iterable
+from typing import Iterator
+from typing import Tuple
+from typing import TypeVar
+
+from scipy.sparse import csr_matrix
+from scipy.sparse import dok_matrix
+from scipy.sparse import identity
+from scipy.sparse import kron
```
Comment (vivalerievich) [inline project/bool_decomposition.py]: import scipy

File: project/bool_decomposition.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+__all__ = ["BoolDecomposition"]
+
+T = TypeVar("T")
+
+
+class BoolDecomposition(Generic[T]):
+    matrices: Dict[T, csr_matrix]
+    size: int
```
Comment (vivalerievich) [inline project/bool_decomposition.py]: ?

File: project/bool_decomposition.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    def __init__(self, size: int, content: Iterable[Tuple[int, int, T]] = None):
+        self.matrices = dict()
+        self.size = size
+        dok_matrices = dict()
+        if content:
+            for (i, j, elm) in content:
```
Comment (vivalerievich) [inline project/bool_decomposition.py]: elm?

File: project/fa_utils.py [hunk 7]
```diff
@@ -7,7 +7,9 @@
 from pyformlang.finite_automaton import Symbol
 from pyformlang.regular_expression import Regex
 
-__all__ = ["graph_to_nfa", "regex_to_minimal_dfa"]
+__all__ = ["graph_to_nfa", "regex_to_minimal_dfa", "intersect_nfas"]
+
+from project.bool_decomposed_nfa import BoolDecomposedNFA
```
Comment (vivalerievich) [inline project/fa_utils.py]: __all__ должно быть после всех импортов

**Review by vadyushkins — CHANGES_REQUESTED** [review]

File: project/bool_decomposed_nfa.py (hunk 1 above)
Comment (IlyaMuravjov) [reply project/bool_decomposed_nfa.py]: Заменил, но очень громоздко получается.

File: project/bool_decomposed_nfa.py (hunk 2 above)
Comment (IlyaMuravjov) [reply project/bool_decomposed_nfa.py]: Убрал. Не знал, что Python так работает, не приходилось им до этого пользоваться. Спасибо, что обратили на это внимание.

File: project/bool_decomposed_nfa.py (hunk 3 above)
Comment (IlyaMuravjov) [reply project/bool_decomposed_nfa.py:46]: Если вопрос про само поле `idx_to_state`, то оно нужно, чтобы после того, как всё отработает, вернуть, не то, что, например, узел, который мы занумеровали третьим, достижим из пятого, а то, какой именно узел достижим из какого именно узла. 

А если вопрос про `__getitem__`, то оно здесь, потому что при создании `BoolDecomposedNFA` для пересечения автоматов туда записывается более сложна функция:
```python
idx_to_state=lambda i: pyformlang.finite_automaton.State(
                     (
                           self.idx_to_state(i // other.state_count),
                           (other.idx_to_state(i % other.state_count)),
                     )
            )
```

File: project/bool_decomposition.py (hunk 4 above)
Comment (IlyaMuravjov) [reply project/bool_decomposition.py]: Заменил

File: project/bool_decomposition.py (hunk 5 above)
Comment (IlyaMuravjov) [reply project/bool_decomposition.py]: Убрал

File: project/bool_decomposition.py (hunk 6 above)
Comment (IlyaMuravjov) [reply project/bool_decomposition.py]: Заменил на `element`. Если вопрос был: "Почему не `label`?", то это, потому что я специально сделал более общую абстракцию для булевой декомпозиции произвольной матрицы, элементы которой -- множества (необязательно для матрицы смежности конечного автомата). Так как, судя по тому, что у нас на лекциях, такая конструкция довольно часто используется.

File: project/fa_utils.py (hunk 7 above)
Comment (IlyaMuravjov) [reply project/fa_utils.py]: Исправил

**Comment by IlyaMuravjov** [comment]
Отключил тест `get_graph_data`, так как `cfpq_data` теперь падает при попытки скачать граф.

**Review by vadyushkins — APPROVED** [review]
в целом норм, зачтено
поправьте пожалуйста тесты - один json на один тест (функция \ метод)
